sympy/integrals/integrals.py

```diff
@@ -1242,6 +1242,105 @@ def _sage_(self):
+        oo
+
+        >>> f = 1 / (x**3)
+
+        >>> Integral(f, (x, -oo, oo)).principal_value()
+        0
+        >>> Integral(f, (x, -10, 10)).principal_value()
+        0
+        >>> Integral(f, (x, -10, oo)).principal_value() + Integral(f, (x, -oo, 10)).principal_value()
+        0
+
+        References
+        ==========
+
+        .. [1] http://en.wikipedia.org/wiki/Cauchy_principal_value
+        .. [2] http://mathworld.wolfram.com/CauchyPrincipalValue.html
+        """
+        from sympy.calculus import singularities
+
+        if len(self.limits) != 1 or len(list(self.limits[0])) != 3:
+            raise ValueError("You need to insert a variable, lower_limit, and upper_limit correctly to calculate "
+                             "cauchy's principal value")
+
+        x, a, b = self.limits[0]
+
+        if not(a.is_comparable and b.is_comparable and a <= b):
+            raise ValueError("The lower_limit must be smaller than or equal to the upper_limit to calculate "
+                             "cauchy's principal value. Also, a and b need to be comparable.")
+
+        if a == b:
+            return 0
+        r = Dummy('r')
+        u = Dummy('u')
+
+        singularities_list = [s for s in singularities(self.function, x) if s.is_comparable and a <= s <= b]
+
+        for i in singularities_list:
+            if ((i == b) or (i == a)):
+                raise ValueError(
+                    'The principal value is not defined in the given interval due to singularity at %d.' % (i))
+
+        f = self.function
+        F = integrate(f, x)
+
+        if F.has(Integral):
+            return self
+        if len(singularities_list) == 0:
+            if a is -oo and b is oo:
+                I = limit(F - F.subs(x, -x), x, oo)
+                return I
+            else:
+                I = limit(F, x, b, '-') - limit(F, x, a, '+')
+                return I
+
+        if len(singularities_list) == 1 and not(a is -oo and b is oo):
+            I = limit(((F.subs(x, singularities_list[0] - r)) - F.subs(x, singularities_list[0] + r)), r, 0,
+                      '+') - limit(F, x, a, '+') + limit(F, x, b, '-')
+            return I
+
+
+        else:
+            if a is -oo and b is oo:
+                Iinf = 0
+                I_lim_subs = limit(F.subs(x, u) - F.subs(x, -u), u, oo)
+                for singular_element in range(len(singularities_list)):
+                    Iinf = Iinf + limit(((F.subs(x, singularities_list[singular_element] - r)) - F.subs(x,
+                                                                                                        singularities_list[
+                                                                                                            singular_element] + r)),
+                                        r, 0, '+')
+                return Iinf + I_lim_subs
+            else:
+                Ib = 0
+                Ia = F.subs(x, singularities_list[0] - r) - limit(F, x, a, '+')
```

Comment:
`F.subs(x, singularities_list[0] - r)` should be handled in the loop.

Reply:
@jksuom I don't see why there's an issue handling it outside the loop.

Reply:
@jksuom Also, that'll destroy the structure of the loop, since the entire `Ia` represents the first integral. So I suggest we keep `F.subs(x, singularities_list[0] - r)` outside the `for` loop

Reply:
Why would you want to handle `singularities_list[0]` differently from other singularities?

Reply:
It will only make the code more complicated.

Reply:
> the entire Ia represents the first integral.

We don't know that `singularities_list[0]` would necessarily be the least singularity. And there is no need to use special names like `Ia` for integrals over various subintervals.

Reply:
@jksuom  Don't you think so, Sir?